### DISCUSSION TITLE
datatypes, counts, and displacements

### DISCUSSION THREAD

**Discussion by lurobi** [mpi4py/mpi4py#120] [Q&A] [answered]
I'm new to MPI, but trying to jump on to a rather ambitious project (for me):  In my simulation, I would like to first divide a matrix across one axis, do some processing, then shuffle it around so that it is divided across the other axis.  Rather than send all data to rank=0, for redistribution, or execute a bunch of send and receives, I think I can do this with Alltoallw, using datatypes of vectors and displacements.

My issue is I seem to have a hard time avoiding MPI_ERR_TRUNCATE which I take as meaning my receive buffer is not correctly sized relative to my sends.  I've reduced a test case down to one send/receive on a single node, which seem to be less picky about the truncation error:

Test Case:
```python
opt = 'vector'
send_arr = np.array([ [r+c/10 for c in range(10)] for r in range(5)])
rcv_arr = np.zeros([5,10])

comm.Send( (send_arr, 50, MPI.DOUBLE), dest=0 )
if opt == 'basic':
    comm.Recv( (rcv_arr, 50, MPI.DOUBLE), source=0 )
elif opt == 'vector':
    vtype = MPI.DOUBLE.Create_vector( 5, 10, 0).Commit()
    comm.Recv( (rcv_arr, 1, vtype), source=0 )

print(f"Send-Recv {opt}: got\n {rcv_arr}")
```

When I run this, with opt = basic, I get my matrix as expected:
```
Send-Recv basic: got
[[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9]
 [1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9]
 [2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9]
 [3.  3.1 3.2 3.3 3.4 3.5 3.6 3.7 3.8 3.9]
 [4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9]]
```
But for opt = vector, I get only a portion of the matrix:
```
Send-Recv vector: got
[[4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9]
 [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
 [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
 [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
 [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]]
```

I was intending the above code to receive 1 block of data with a type defined as `Create_vector( 5, 10, 0)` which I interpreted as 5 blocks of 10 adjacent  elements with 0 elements of stride between blocks for a total of 50 elements.  Based on these results, I'm clearly not understanding that correctly.

**Comment by dalcinl** [comment, marked as answer]
On Tue, 19 Oct 2021 at 23:37, lurobi ***@***.***> wrote:

> I'm new to MPI, but trying to jump on to a rather ambitious project (for
> me):
>
Ambition is the fuel of progress.

And yes, Alltoallw and datatypes can greatly simplify your code.

Test Case:
>
Your testcase is not complete. Is comm = MPI.COMM_SELF? Or are you running
on with a single MPI process (which is not the same as running on a single
node, where you can still have many MPI processes).

opt = 'vector'send_arr = np.array([ [r+c/10 for c in range(10)] for r
in range(5)])rcv_arr = np.zeros([5,10])
> comm.Send( (send_arr, 50, MPI.DOUBLE), dest=0 )if opt == 'basic':
>     comm.Recv( (rcv_arr, 50, MPI.DOUBLE), source=0 )elif opt == 'vector':
>     vtype = MPI.DOUBLE.Create_vector( 5, 10, 0).Commit()
>     comm.Recv( (rcv_arr, 1, vtype), source=0 )
>
>
This is a send-to-self, your code is non-portable, it can deadlock if MPI
does not internally buffer the send message.
You have to use req = comm.Isend(...), and after the recv call, do
req.Wait()

> print(f"Send-Recv {opt}: got\n {rcv_arr}")
>
> When I run this, with opt = basic, I get my matrix as expected:
>
> Send-Recv basic: got
> [[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9]
>  [1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9]
>  [2.  2.1 2.2 2.3 2.4 2.5 2.6 2.7 2.8 2.9]
>  [3.  3.1 3.2 3.3 3.4 3.5 3.6 3.7 3.8 3.9]
>  [4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9]]
>
> But for opt = vector, I get only a portion of the matrix:
>
> Send-Recv vector: got
> [[4.  4.1 4.2 4.3 4.4 4.5 4.6 4.7 4.8 4.9]
>  [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
>  [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
>  [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
>  [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]]
>
> I was intending the above code to receive 1 block of data with a type
> defined as Create_vector( 5, 10, 0) which I interpreted as 5 blocks of 10
> adjacent elements with 0 elements of stride between blocks for a total of
> 50 elements. Based on these results, I'm clearly not understanding that
> correctly.
>

Stride cannot be 0. You have to use stride=10 if you want the "vector"
version to be equal to the "basic" version.


-- 
Lisandro Dalcin
============
Senior Research Scientist
Extreme Computing Research Center (ECRC)
King Abdullah University of Science and Technology (KAUST)
http://ecrc.kaust.edu.sa/

**Reply by lurobi** [reply]
Thank you for the tip on strides,  I had interpreted the stride as elements between end and next beginning but based on your comment, I should interpret it as elements between subsequent beginnings.  I think this was my major mix-up.

With that in mind, I fixed the remaining logic, and am now able to send different parts of the matrix to different ranks.  Some other things I think I've learned along the way:

 * I can send a basic data type (DOUBLE), and receive a vector of that datatype (as I had hoped).
 * When receiving a vector which writes to every other block of 10 elements, I should use `vector(5,10,20)`, and I should receive exactly ONE such entity of my defined type.  Trying to receive FIVE smaller vectors as `vector(1,10,20)` does not respect the stride between subsequent datatypes, and so doesn't work for that every-other block of 10 case. 

Thanks for the help Lisandro!

**Comment by chakpongchung** [comment]
could you provide a `alltoallw()` example for `set()`?

For example, I want to do alltoall for the following list between all processes:

```
before:
rank: 0, senddata: [{2}, {5, 7}], recvdata: [set(), set()]
rank: 1, senddata: [set(), set()], recvdata: [set(), set()]

after:
rank: 0, senddata: [{2}, {5, 7}], recvdata: [{2}, set()]
rank: 1, senddata: [set(), set()], recvdata: [set(), {5,7}]
```

Currently this does not work:

```
import numpy
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()
# print(size)
print(rank)
a_size = 1
senddata = None
recvdata = None
if rank == 0:
    senddata = [{2}, {5, 7}]
    recvdata = [set(), set()]
elif rank == 1:
    senddata = [set(), set()]
    recvdata = [set(), set()]
else:
    exit("did you run mpiexec -n 2 python alltoall_example.py  ?")

# This input works for alltoall() and alltoallv()
# senddata = (rank + 1) * numpy.arange(size, dtype=int)
# recvdata = numpy.zeros(size * a_size, dtype=int)
print("before rank %s senddata %s recvdata %s " % (rank, senddata, recvdata))
comm.Alltoallw(senddata, recvdata)
print("aftr   rank %s senddata %s recvdata %s " % (rank, senddata, recvdata))

# expectation:
# before:
# rank: 0, senddata: [{2}, {5, 7}], recvdata: [set(), set()]
# rank: 1, senddata: [set(), set()], recvdata: [set(), set()]
#
# after:
# rank: 0, senddata: [{2}, {5, 7}], recvdata: [{2}, set()]
# rank: 1, senddata: [set(), set()], recvdata: [set(), {5,7}]

# mpiexec -n 2 python alltoall_example.py
```

**Reply by dalcinl** [reply]
You cannot do that with `Alltoallw()`, uppercase methods only support buffer-like objects like NumPy arrays. If you want to exchange arbitrary Python object in an all-to-all fashion, you just need to do:

```
recvdata = comm.alltoall(senddata)  # note I'm using lowercase `alltoall()`, and the received data is in the result.
```

Replace your `Alltoallw()` call with the line above and you should get what you are looking after.

**Reply by chakpongchung** [reply]
lets move the discussion here:
https://github.com/mpi4py/mpi4py/discussions/136
